spec/controllers/links_controller_spec.rb

```diff
@@ -1,69 +1,161 @@
-require 'rails_helper'
+require "rails_helper"
 
 RSpec.describe LinksController, type: :controller do
+  let(:user) { FactoryGirl.create(:user) }
 
-  # This should return the minimal set of attributes required to create a valid
-  # Link. As you add validations to Link, be sure to
-  # adjust the attributes here as well.
-  let(:valid_attributes) { 
-    {:original => "http://something.com", :user_id => 2}
-  }
-
-  let(:invalid_attributes) {
-    {:original => "something.com", :user_id => 2}
-    {:original => "http://something.com", :user_id => nil}
-  }
-
-  # This should return the minimal set of values that should be in the session
-  # in order to pass any filters (e.g. authentication) defined in
-  # LinksController. Be sure to keep this updated too.
-  # let(:valid_session) { {} }
-
-  describe "GET #show" do
-    include RSpec::Rails::RequestExampleGroup
-    it "redirects slug to original url" do
-      link = Link.create! valid_attributes
-      get link.shortened_url
-      expect(response).to redirect_to(link.original)
+  let(:link) { Link.create(FactoryGirl.attributes_for(:link)) }
+
+  describe "before actions" do
+    it { is_expected.to use_before_action(:authenticate) }
+    it { is_expected.to use_before_action(:set_link) }
+    it { is_expected.to use_before_action(:my_links) }
+  end
+
+  describe "#show" do
+    context "when slug is valid and active" do
+      it "redirects user to original url" do
+        get :show, slug: link.slug
+
+        expect(response).to redirect_to(link.original)
+        expect(Visit.count).to eq 1
+      end
+    end
+
+    context "when slug is inactive" do
+      it "redirects user to original url" do
+        link.update(slug: "inactive", active: false)
+        get :show, slug: "inactive"
+
+        expect(response).to render_template "layouts/error"
+      end
+    end
+
+    context "when slug has been deleted" do
+      it "redirects user to original url" do
+        get :show, slug: "deleted"
+
+        expect(response).to render_template "layouts/error"
+      end
     end
   end
 
-  describe "GET #new" do
-    it "assigns a new link as @link" do
-      get :new, params: {}
-      expect(assigns(:link)).to be_a_new(Link)
+  describe "#dashboard" do
+    context "when user is registered" do
+      it "renders the dashboard" do
+        session[:user_id] = user.id
+
+        get :dashboard
+
+        expect(response).to render_template "dashboard"
+      end
+    end
+
+    context "when user is not registered" do
+      it "asks user to login" do
+        get :dashboard
+
+        expect(response).to redirect_to login_path
+      end
+    end
+  end
+
+  describe "#new" do
+    it "shows form to create a shortened link" do
+      get :new
+
+      expect(response).to render_template "new"
     end
   end
 
-  describe "POST #create" do
+  describe "#edit" do
+    context "when user is not registered" do
+      it "asks user to login" do
```

Comment:
**Critical - Test Documentation**
Rather, it `redirects user to the login page`.

You always want your descriptions to be in line with what's actually happening and/or is being tested.

Please fix all instances of this issue.